File: src/bun.js/api/Timer/TimeoutObject.zig
const Self = @This();

const RefCount = bun.ptr.RefCount(@This(), "ref_count", deinit, .{});
pub const ref = RefCount.ref;
pub const deref = RefCount.deref;

pub const js = jsc.Codegen.JSTimeout;
pub const toJS = js.toJS;
pub const fromJS = js.fromJS;
pub const fromJSDirect = js.fromJSDirect;

ref_count: RefCount,
event_loop_timer: EventLoopTimer = .{
    .next = .epoch,
    .tag = .TimeoutObject,
},
internals: TimerObjectInternals,

pub fn init(
    globalThis: *JSGlobalObject,
    id: i32,
    kind: Kind,
    interval: u31,
    callback: JSValue,
    arguments: JSValue,
) JSValue {
    // internals are initialized by init()
    const timeout = bun.new(Self, .{ .ref_count = .init(), .internals = undefined });
    const js_value = timeout.toJS(globalThis);
    defer js_value.ensureStillAlive();
    timeout.internals.init(
        js_value,
        globalThis,
        id,
        kind,
        interval,
        callback,
        arguments,
    );

    if (globalThis.bunVM().isInspectorEnabled()) {
        Debugger.didScheduleAsyncCall(
            globalThis,
            .DOMTimer,
            ID.asyncID(.{ .id = id, .kind = kind.big() }),
            kind != .setInterval,
        );
    }

    return js_value;
}

fn deinit(self: *Self) void {
    self.internals.deinit();
    bun.destroy(self);
}

pub fn constructor(globalObject: *jsc.JSGlobalObject, callFrame: *jsc.CallFrame) !*Self {
    _ = callFrame;
    return globalObject.throw("Timeout is not constructible", .{});
}

pub fn toPrimitive(self: *Self, _: *JSGlobalObject, _: *jsc.CallFrame) bun.JSError!JSValue {
    return self.internals.toPrimitive();
}

pub fn doRef(self: *Self, globalThis: *JSGlobalObject, callFrame: *jsc.CallFrame) bun.JSError!JSValue {
    return self.internals.doRef(globalThis, callFrame.this());
}

pub fn doUnref(self: *Self, globalThis: *JSGlobalObject, callFrame: *jsc.CallFrame) bun.JSError!JSValue {
    return self.internals.doUnref(globalThis, callFrame.this());
}

pub fn doRefresh(self: *Self, globalThis: *JSGlobalObject, callFrame: *jsc.CallFrame) bun.JSError!JSValue {
    return self.internals.doRefresh(globalThis, callFrame.this());
}

pub fn hasRef(self: *Self, _: *JSGlobalObject, _: *jsc.CallFrame) bun.JSError!JSValue {
    return self.internals.hasRef();
}

pub fn finalize(self: *Self) void {
    self.internals.finalize();
}

pub fn getDestroyed(self: *Self, globalThis: *JSGlobalObject) JSValue {
    _ = globalThis;
    return .jsBoolean(self.internals.getDestroyed());
}

pub fn close(self: *Self, globalThis: *JSGlobalObject, callFrame: *jsc.CallFrame) JSValue {
    self.internals.cancel(globalThis.bunVM());
    return callFrame.this();
}

pub fn get_onTimeout(_: *Self, thisValue: JSValue, _: *JSGlobalObject) JSValue {
    return Self.js.callbackGetCached(thisValue).?;
}

pub fn set_onTimeout(_: *Self, thisValue: JSValue, globalThis: *JSGlobalObject, value: JSValue) void {
    Self.js.callbackSetCached(thisValue, globalThis, value);
}

pub fn get_idleTimeout(_: *Self, thisValue: JSValue, _: *JSGlobalObject) JSValue {
    return Self.js.idleTimeoutGetCached(thisValue).?;
}

pub fn set_idleTimeout(_: *Self, thisValue: JSValue, globalThis: *JSGlobalObject, value: JSValue) void {
    Self.js.idleTimeoutSetCached(thisValue, globalThis, value);
}

pub fn get_repeat(_: *Self, thisValue: JSValue, _: *JSGlobalObject) JSValue {
    return Self.js.repeatGetCached(thisValue).?;
}

pub fn set_repeat(_: *Self, thisValue: JSValue, globalThis: *JSGlobalObject, value: JSValue) void {
    Self.js.repeatSetCached(thisValue, globalThis, value);
}

pub fn get_idleStart(_: *Self, thisValue: JSValue, _: *JSGlobalObject) JSValue {
    return Self.js.idleStartGetCached(thisValue).?;
}

pub fn set_idleStart(_: *Self, thisValue: JSValue, globalThis: *JSGlobalObject, value: JSValue) void {
    Self.js.idleStartSetCached(thisValue, globalThis, value);
}

pub fn dispose(self: *Self, globalThis: *JSGlobalObject, _: *jsc.CallFrame) bun.JSError!JSValue {
    self.internals.cancel(globalThis.bunVM());
    return .js_undefined;
}

const Debugger = @import("../../Debugger.zig");
const bun = @import("bun");

const EventLoopTimer = bun.api.Timer.EventLoopTimer;
const ID = bun.api.Timer.ID;
const Kind = bun.api.Timer.Kind;
const TimerObjectInternals = bun.api.Timer.TimerObjectInternals;

const jsc = bun.jsc;
const JSGlobalObject = jsc.JSGlobalObject;
const JSValue = jsc.JSValue;

File: src/runtime/timer/TimeoutObject.rs
use bun_jsc::generated::JSTimeout as js;
use bun_jsc::{CallFrame, JSGlobalObject, JSValue, JsResult};

use super::Kind;

// Struct + `RefCounted`/`Default` impls + the forwarder host-fns
// (`to_primitive`/`do_ref`/`do_unref`/`has_ref`/`get_destroyed`/`dispose`/
// `constructor`/`finalize`/`ref_`/`deref`/`deinit`/`init_with`) — see
// `impl_timer_object!` in `super` (timer/mod.rs).
super::impl_timer_object!(TimeoutObject, TimeoutObject, "Timeout");

impl TimeoutObject {
    pub(crate) fn init(
        global: &JSGlobalObject,
        id: i32,
        kind: Kind,
        interval: u32,
        callback: JSValue,
        arguments: JSValue,
    ) -> JSValue {
        Self::init_with(global, id, kind, interval, callback, arguments)
    }

    #[bun_jsc::host_fn(method)]
    pub(crate) fn do_refresh(
        this: &Self,
        global: &JSGlobalObject,
        frame: &CallFrame,
    ) -> JsResult<JSValue> {
        this.internals.do_refresh(global, frame.this())
    }

    #[bun_jsc::host_fn(method)]
    pub fn close(this: &Self, global: &JSGlobalObject, frame: &CallFrame) -> JsResult<JSValue> {
        this.before_explicit_cancel(global, frame.this());
        this.internals.cancel(global.bun_vm_ptr());
        Ok(frame.this())
    }

    /// Node's `unenroll`: an explicitly cancelled timer reads back `_idleTimeout === -1`.
    pub(crate) fn mark_unenrolled(this_value: JSValue, global: &JSGlobalObject) {
        js::idle_timeout_set_cached(this_value, global, JSValue::js_number(-1.0));
    }

    /// `impl_timer_object!`'s `dispose` hook.
    pub(crate) fn before_explicit_cancel(&self, global: &JSGlobalObject, this_value: JSValue) {
        Self::mark_unenrolled(this_value, global);
    }

    // Cached-property getters/setters — codegen passes `this_value` (the JS
    // wrapper) so the cached `WriteBarrier` slot on the C++ side can be read/written.
    // Signature does not match the standard `host_fn(getter/setter)` shape; the
    // `#[JsClass]` derive emits the C-ABI shims directly.

    pub(crate) fn get_on_timeout(
        _this: &Self,
        this_value: JSValue,
        _global: &JSGlobalObject,
    ) -> JSValue {
        js::callback_get_cached(this_value).unwrap()
    }

    pub(crate) fn set_on_timeout(
        _this: &Self,
        this_value: JSValue,
        global: &JSGlobalObject,
        value: JSValue,
    ) {
        js::callback_set_cached(this_value, global, value);
    }

    pub(crate) fn get_idle_timeout(
        _this: &Self,
        this_value: JSValue,
        _global: &JSGlobalObject,
    ) -> JSValue {
        js::idle_timeout_get_cached(this_value).unwrap()
    }

    pub(crate) fn set_idle_timeout(
        _this: &Self,
        this_value: JSValue,
        global: &JSGlobalObject,
        value: JSValue,
    ) {
        js::idle_timeout_set_cached(this_value, global, value);
    }

    pub(crate) fn get_repeat(
        _this: &Self,
        this_value: JSValue,
        _global: &JSGlobalObject,
    ) -> JSValue {
        js::repeat_get_cached(this_value).unwrap()
    }

    pub(crate) fn set_repeat(
        _this: &Self,
        this_value: JSValue,
        global: &JSGlobalObject,
        value: JSValue,
    ) {
        js::repeat_set_cached(this_value, global, value);
    }

    pub(crate) fn get_idle_start(
        _this: &Self,
        this_value: JSValue,
        _global: &JSGlobalObject,
    ) -> JSValue {
        js::idle_start_get_cached(this_value).unwrap()
    }

    pub(crate) fn set_idle_start(
        this: &Self,
        this_value: JSValue,
        global: &JSGlobalObject,
        value: JSValue,
    ) {
        if let Some(ms) = value.get_number() {
            this.internals.set_idle_start(ms);
        }
        js::idle_start_set_cached(this_value, global, value);
    }
}
